include/nbl/asset/IRenderpass.h

```diff
@@ -13,102 +13,556 @@
+class IRenderpass
+{
+    public:
+        // TODO: move this out somewhere? looks useful
+        template<typename T, class func_t>
+        static inline void visitTokenTerminatedArray(const T* array, const T& endToken, func_t&& func)
         {
-            struct SAttachmentRef
+            if (array)
+            for (auto it=array; *it!=endToken && func(*it); it++)
             {
-                uint32_t attachment = ATTACHMENT_UNUSED;
-                IImage::E_LAYOUT layout = IImage::EL_UNDEFINED;
-
-                auto operator<=>(const SAttachmentRef&) const = default;
-            };
+            }
+        }
 
-            E_SUBPASS_DESCRIPTION_FLAGS flags = ESDF_NONE;
-            E_PIPELINE_BIND_POINT pipelineBindPoint;
-            const SAttachmentRef* depthStencilAttachment;
-            uint32_t inputAttachmentCount;
-            const SAttachmentRef* inputAttachments;
-            //! denotes resolve attachment count as well
-            uint32_t colorAttachmentCount;
-            const SAttachmentRef* colorAttachments;
-            const SAttachmentRef* resolveAttachments;
-            uint32_t preserveAttachmentCount;
-            const uint32_t* preserveAttachments;
-
-            auto operator<=>(const SSubpassDescription&) const = default;
+        enum class E_LOAD_OP : uint8_t
+        {
+            LOAD = 0,
+            CLEAR,
+            DONT_CARE,
+            UNDEFINED
+        };
+        enum class E_STORE_OP: uint8_t
+        {
+            STORE = 0,
+            DONT_CARE,
+            UNDEFINED
+        };
+        enum class E_SUBPASS_DESCRIPTION_FLAGS : uint8_t
+        {
+            NONE = 0x00,
+            PER_VIEW_ATTRIBUTES_BIT = 0x01,
+            PER_VIEW_POSITION_X_ONLY_BIT = 0x02
+            // TODO: VK_EXT_rasterization_order_attachment_access
+        };
+        // For all arrays here we use ArrayNameEnd terminator instead of specifying the count
+        struct SCreationParams
+        {
+            public:
+                // funny little struct to allow us to use ops as `layout.depth`, `layout.stencil` and `layout`
+                struct Layout : SDepthStencilLayout
+                {
+                    inline operator IImage::E_LAYOUT() const { return depth; }
+
+                    auto operator<=>(const Layout&) const = default;
+                };
+                // The reason we don't have separate types per depthstencil, color and resolve is because
+                // attachments limits (1 depth, MaxColorAttachments color and resolve) only apply PER SUBPASS
+                // so overall we can have as many attachments of whatever type and in whatever order we want
+                struct SAttachmentDescription
+                {
+                    // similar idea for load/store ops as for layouts
+                    template<typename op_t>
+                    struct Op
+                    {
+                        op_t depth : 2 = op_t::DONT_CARE;
```

Comment:
I'm a little consufed here, I think we're using `depth` also for color attachment load/store ops and also is `SLayout`?